include/ginkgo/core/base/mpi.hpp

```diff
@@ -78,6 +78,63 @@ static constexpr bool is_gpu_aware()
 }
 
 
+/**
+ * Maps each MPI rank to a single device id in a round robin manner.
+ * @param comm  used to determine the node-local rank, if no suitable
+ *              environment variable is available.
+ * @param num_devices  the number of devices per node.
+ * @return  device id that this rank should use.
+ */
+inline int map_rank_to_device_id(MPI_Comm comm, const int num_devices)
+{
+    GKO_ASSERT(num_devices > 0);
+    if (num_devices == 1) {
+        return 0;
+    } else {
+        auto mpi_node_local_rank = [](MPI_Comm comm_) {
+            int local_rank;
+            MPI_Comm local_comm;
+            GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_split_type(
+                comm_, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &local_comm));
+            GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_rank(local_comm, &local_rank));
+            MPI_Comm_free(&local_comm);
+            return local_rank;
+        };
+
+        // When we are using MPI_COMM_WORLD, there might be already an
+        // environment variable describing the node local rank, so we
+        // prioritize it. If no suitable environment variable is found
+        // we determine the node-local rank with MPI calls.
+        int local_rank;
+        int compare_result;
+        int is_initialized;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Initialized(&is_initialized));
+        if (is_initialized) {
+            GKO_ASSERT_NO_MPI_ERRORS(
+                MPI_Comm_compare(comm, MPI_COMM_WORLD, &compare_result));
+        } else {
+            compare_result = MPI_IDENT;
+        }
+        if (compare_result != MPI_IDENT && compare_result != MPI_CONGRUENT) {
+            local_rank = mpi_node_local_rank(comm);
+        } else {
+            if (auto str = std::getenv("MV2_COMM_WORLD_LOCAL_RANK")) {
+                local_rank = std::stoi(str);
+            } else if (auto str = std::getenv("OMPI_COMM_WORLD_LOCAL_RANK")) {
+                local_rank = std::stoi(str);
+            } else if (auto str = std::getenv("MPI_LOCALRANKID")) {
+                local_rank = std::stoi(str);
+            } else if (auto str = std::getenv("SLURM_LOCALID")) {
+                local_rank = std::stoi(str);
+            } else {
+                local_rank = mpi_node_local_rank(comm);
+            }
```

Comment:
Yes, that's true, and that's why I've added the default case. So if we can't determine the local rank from environment variables (which is think is preferable) then we use the open mpi construct for that.

Reply:
I feel this this might create behaviour that the user not intend or want. For example, Summit uses variables from BSUB i think. So, it user sets it using those variables, then we just disregard them. In general, I think taking into account user settings from different places is a hard problem, and I dont have a solution right now. But that might be something we have to think about when interfacing to other applications.

Reply:
In your Summit example, we would not be able to handle it anyway, so I guess nothing is lost.
For interfacing external libraries, if they have MPI already set up, they can just pass us in the local rank. I guess there needs to be a nice way to insert that information.